File: Ejercicios_etapa_1/Lua/Ejercicio_7/prueba.test.lua
-- oop.test.lua
package.path = "./?.lua;" .. package.path
local oop = require("oop")

describe("Patrón Factory - Usuario", function()
    it("debe crear un usuario y retornar el saludo correcto", function()
        local usuario = oop.crearUsuarioConMetatable("Luis", "admin")
        local saludo = usuario:saludar()
        assert.are.equal("Hola, soy Luis y soy admin", saludo)
    end)
end)

describe("Patrón PubSub", function()
    it("debe notificar a los suscriptores cuando se publica un evento", function()
        local pubsub = oop.PublicadorSubscriptor:new()
        local mensajeRecibido = nil
        local suscripcion = pubsub:suscribirse(oop.PublicadorSubscriptor.TEMAS.MOSTRAR_NOTIFICACION_EXITO, function(info)
            mensajeRecibido = info.mensaje
        end)
        pubsub:publicar(oop.PublicadorSubscriptor.TEMAS.MOSTRAR_NOTIFICACION_EXITO, { mensaje = "¡Hola mundo!" })
        assert.are.equal("¡Hola mundo!", mensajeRecibido)
    end)

    it("debe cancelar la suscripción correctamente", function()
        local pubsub = oop.PublicadorSubscriptor:new()
        local contador = 0
        local suscripcion = pubsub:suscribirse(oop.PublicadorSubscriptor.TEMAS.MOSTRAR_NOTIFICACION_EXITO, function(info)
            contador = contador + 1
        end)
        pubsub:publicar(oop.PublicadorSubscriptor.TEMAS.MOSTRAR_NOTIFICACION_EXITO, { mensaje = "Evento 1" })
        suscripcion.remover()
        pubsub:publicar(oop.PublicadorSubscriptor.TEMAS.MOSTRAR_NOTIFICACION_EXITO, { mensaje = "Evento 2" })
        assert.are.equal(1, contador)
    end)
end)

File: Ejercicios_etapa_1/Javascript/Ejercicio_7/prueba.test.js
// prueba.test.js

import { describe, it, expect, vi } from 'vitest';
import { crearUsuarioConPrototipo, Usuario, PubSub } from './index.js';

describe('crearUsuarioConPrototipo', () => {
  it('debería crear un objeto con nombre, rol y método saludar', () => {
    const usuario = crearUsuarioConPrototipo("Carlos", "editor");
    expect(usuario).toBeTypeOf("object");
    expect(usuario.nombre).toBe("Carlos");
    expect(usuario.rol).toBe("editor");
    expect(usuario.saludar()).toBe("Hola, soy Carlos y soy editor.");
  });
});

describe('Usuario (Clase)', () => {
  it('debería crear una instancia de Usuario con nombre, rol y método saludar', () => {
    const usuario = new Usuario("Lucía", "moderadora");
    expect(usuario).toBeInstanceOf(Usuario);
    expect(usuario.nombre).toBe("Lucía");
    expect(usuario.rol).toBe("moderadora");
    expect(usuario.saludar()).toBe("Hola, soy Lucía y soy moderadora.");
  });
});

describe('PubSub', () => {
  it('debería permitir suscribirse y recibir una publicación', () => {
    const pubsub = new PubSub();
    const fn = vi.fn();

    pubsub.suscribirse("NOTIFICACION", fn);
    pubsub.publicar("NOTIFICACION", { mensaje: "¡Funciona!" });

    expect(fn).toHaveBeenCalledOnce();
    expect(fn).toHaveBeenCalledWith({ mensaje: "¡Funciona!" });
  });

  it('debería permitir remover una suscripción', () => {
    const pubsub = new PubSub();
    const fn = vi.fn();

    const { remover } = pubsub.suscribirse("ALERTA", fn);
    remover(); // eliminamos el listener
    pubsub.publicar("ALERTA", { mensaje: "Esto no debe recibirse" });

    expect(fn).not.toHaveBeenCalled();
  });

  it('no debería fallar si se publica un evento sin suscriptores', () => {
    const pubsub = new PubSub();
    expect(() => pubsub.publicar("VACIO", { mensaje: "sin listeners" })).not.toThrow();
  });
});
